src/acceleration/BaseQNAcceleration.cpp

```diff
@@ -170,7 +174,100 @@ void BaseQNAcceleration::initialize(
 
   _preconditioner->initialize(subVectorSizes);
 }
+void BaseQNAcceleration::projectFWInput(DataMap &cplData, const std::vector<DataID> &dataIDs, std::map<int, double> lowerBounds,
+                                        std::map<int, double> upperBounds)
+{
+  Eigen::Index offset = 0;
+  for (auto id : dataIDs) {
+    Eigen::Index size       = cplData.at(id)->values().size();
+    auto         lowerBound = lowerBounds.at(id);
+    auto         upperBound = upperBounds.at(id);
+    if (lowerBound <= -1.0e16 && upperBound >= 1.0e16) { // do nothing
+    } else if (lowerBound > -1.0e16 && upperBound < 1.0e16) {
+      // when both sides are bounded
+      double delta       = 0.1;
+      double leftLimit   = lowerBound - delta;
+      double rightLimit  = upperBound + delta;
+      double intervalLen = rightLimit - leftLimit;
+      // double factor        = 1;
+      // int selectedTrans = 1; // 1: logTrans, 2: tanhTrans, else: noTrans
+
+      for (Eigen::Index i = 0; i < size; i++) {
+        double normOldValue = (_oldValues[i + offset] - leftLimit) / intervalLen;
+        double normValue    = (_values[i + offset] - leftLimit) / intervalLen;
```

Comment:
```suggestion
        double normalizedOldValue = (_oldValues[i + offset] - leftLimit) / intervalLen;
        double normalizedValue    = (_values[i + offset] - leftLimit) / intervalLen;
```
Otherwise, it sounds like a norm.